File: backend/src/salora_platform/observability.py
"""Structured privacy-safe logs and in-process metrics. Does not replace stdlib loggers."""

from __future__ import annotations

import json
import logging
import re
import time
from typing import Any

FORBIDDEN = re.compile(
    r"transcript|utterance|otp|phone|secret|password|token|api[_-]?key|prompt",
    re.IGNORECASE,
)

_counts: dict[str, float] = {}


def _redact(fields: dict[str, Any]) -> dict[str, Any]:
    clean: dict[str, Any] = {}
    for key, value in fields.items():
        if FORBIDDEN.search(key):
            continue
        if isinstance(value, str) and FORBIDDEN.search(value) and len(value) > 24:
            continue
        clean[key] = value
    return clean


def emit(level: str, event: str, **fields: Any) -> None:
    line = {
        "ts": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "level": level,
        "event": event,
        **_redact(fields),
    }
    logging.getLogger("salora.platform").log(
        getattr(logging, level.upper(), logging.INFO),
        json.dumps(line, default=str),
    )


def record_metric(name: str, value: float = 1.0) -> None:
    _counts[name] = _counts.get(name, 0.0) + value


def metric_snapshot() -> dict[str, float]:
    return dict(_counts)


def reset_metrics() -> None:
    _counts.clear()


def heartbeat(service: str = "backend") -> None:
    record_metric("heartbeat")
    emit("info", "heartbeat", service=service)

File: frontend/lib/platform/observability.ts
/**
 * Structured logs, metrics, traces. Privacy-safe. No speech, OTP, phone, secrets.
 */

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

export type MetricName =
  | 'voice.session.start'
  | 'voice.session.end'
  | 'voice.latency.connect_ms'
  | 'voice.latency.first_audio_ms'
  | 'api.request'
  | 'api.latency_ms'
  | 'api.error'
  | 'learning.projection'
  | 'adaptive.decision'
  | 'knowledge.retrieval'
  | 'agent.handoff'
  | 'agent.handback'
  | 'heartbeat';

export type TraceSpan = {
  name: string;
  startMs: number;
  endMs?: number;
  ok?: boolean;
};

const FORBIDDEN = /transcript|utterance|otp|phone|secret|password|token|api[_-]?key|prompt/i;

type LogFields = Record<string, unknown>;

type MetricPoint = {
  name: MetricName;
  value: number;
  at: number;
  tags?: Record<string, string>;
};

const metrics: MetricPoint[] = [];
const MAX_POINTS = 500;

function redact(fields: LogFields): LogFields {
  const clean: LogFields = {};
  for (const [key, value] of Object.entries(fields)) {
    if (FORBIDDEN.test(key)) continue;
    if (typeof value === 'string' && FORBIDDEN.test(value) && value.length > 24) continue;
    clean[key] = value;
  }
  return clean;
}

export function log(level: LogLevel, event: string, fields: LogFields = {}): void {
  const line = {
    ts: new Date().toISOString(),
    level,
    event,
    ...redact(fields),
  };
  const serialized = JSON.stringify(line);
  if (level === 'error') console.error(serialized);
  else if (level === 'warn') console.warn(serialized);
  else console.info(serialized);
}

export function recordMetric(name: MetricName, value = 1, tags?: Record<string, string>): void {
  metrics.push({ name, value, at: Date.now(), tags });
  if (metrics.length > MAX_POINTS) metrics.splice(0, metrics.length - MAX_POINTS);
}

export function getMetricsSnapshot(): {
  counts: Record<string, number>;
  last: MetricPoint[];
} {
  const counts: Record<string, number> = {};
  for (const point of metrics) {
    counts[point.name] = (counts[point.name] ?? 0) + point.value;
  }
  return { counts, last: metrics.slice(-50) };
}

export function resetMetrics(): void {
  metrics.length = 0;
}

export function startSpan(name: string): TraceSpan {
  return { name, startMs: Date.now() };
}

export function endSpan(span: TraceSpan, ok = true): number {
  span.endMs = Date.now();
  span.ok = ok;
  const duration = span.endMs - span.startMs;
  log('info', 'span.end', { span: span.name, duration_ms: duration, ok });
  return duration;
}

export function heartbeat(service: string): void {
  recordMetric('heartbeat', 1, { service });
  log('info', 'heartbeat', { service });
}
